File: old_contents/boj_codes/Python3/9935.py
# https://www.acmicpc.net/problem/9935
# 20201215

import sys

class Stack:
    def __init__(self):
        self.list = [0] * 1000001
        self.state_list = [0] * 1000001
        self.top = -1

    def insert(self, word):
        self.top += 1
        self.list[self.top] = word

    def pops(self, n_pop):
        self.top -= n_pop

    def get_results(self):
        return self.list[:self.top + 1]

    def set_state(self, state):
        self.state_list[self.top] = state

    def pop_state(self):
        if self.top == -1:
            return 0
        else:
            return self.state_list[self.top]


if __name__ == "__main__":
    words = sys.stdin.readline().rstrip()
    bomb = sys.stdin.readline().rstrip()
    len_words = len(words)
    len_bomb = len(bomb)

    # if cur_match == len(bomb) : bomb it!
    cur_match = 0

    stack = Stack()

    for cur_word in words:
        stack.insert(cur_word)


        if cur_word == bomb[cur_match]:
            cur_match += 1
            stack.set_state(cur_match)

            if cur_match == len_bomb:
                stack.pops(len_bomb)
                cur_match = stack.pop_state()

        else:
            cur_match = 0

            if cur_word == bomb[cur_match]:
                cur_match += 1
                stack.set_state(cur_match)
            else:
                stack.set_state(0)

    res = stack.get_results()
    if not res:
        print("FRULA")

    else:
        print(''.join(res))

File: old_contents/boj_codes/C++17/9935.cpp
// https://www.acmicpc.net/problem/9935
// First Written : 20210118
// Last Modified : 20210118

#include <cstdio>
#include <cstring>
#include <vector>

#define MAX_LEN 1000001
#define MAX_BOMB_LEN 37

using namespace std;

class Stack{
private:
    vector<char> list;
    vector<int> state_list;
public:

    void insert(char word){
        list.push_back(word);
        state_list.push_back(0);
    }

    void set_state(int state){
        state_list.pop_back();
        state_list.push_back(state);
    }

    void pops(int n_pop){
        while(n_pop--){
            list.pop_back();
            state_list.pop_back();
        }
    }

    int get_state(){
        if(state_list.size() == 0)
            return 0;
        
        else
            return state_list[state_list.size() - 1];
    }

    int get_size(){
        return list.size();
    }

    void print_result(){
        for(auto it:list)
            printf("%c", it);
    }

};

char original[MAX_LEN];
char bomb[MAX_BOMB_LEN];

int main(){
    scanf("%s", original);
    scanf("%s", bomb);

    int original_len = strlen(original);
    int bomb_len = strlen(bomb);
    int cur_match = 0;
    char cur_word;

    Stack stack;

    for(int i = 0; i < original_len; i++){
        cur_word = original[i];
        stack.insert(cur_word);

        if(cur_word == bomb[cur_match]){
            stack.set_state(++cur_match);

            if(cur_match == bomb_len){
                stack.pops(bomb_len);
                cur_match = stack.get_state();
            }
        }

        else{
            cur_match = 0;

            if(cur_word == bomb[cur_match]){
                stack.set_state(++cur_match);
            }
        }


    }

    if(stack.get_size() == 0)
        printf("FRULA");
    
    else
        stack.print_result();
    
    return 0;
}